File: Cf/186/C.java
import java.util.*;
import java.math.*;
import java.io.*;

public class C {

    public static void main(String[] args) throws Exception {
    	BufferedReader buffer = new BufferedReader(new InputStreamReader(System.in));
    	int n = Integer.parseInt(buffer.readLine());
    	String[] str = buffer.readLine().split(" ");
    	PriorityQueue<Integer> pq = new PriorityQueue<Integer>(1, 
    		new Comparator<Integer>() {
    			public int compare(Integer a, Integer b) {
    				return b-a;
    			}
    		}
    	);
    	BigInteger ans = BigInteger.ZERO;
    	for(int i=0; i<n; i++) {
    		int tmp = Integer.parseInt(str[i]);
    		pq.offer(tmp);
			ans = ans.add(new BigInteger("" + tmp));
    	}
    	int x = ((int)(Math.log(n)/Math.log(4)));
    	int j=0;
    	for(int i=n-1; i>=Math.max(0, n-(int)(Math.pow(4, x-1))); i--) {
    		int tmp = pq.poll();
    		ans = ans.add(new BigInteger(""+tmp).multiply(new BigInteger(""+(x-(int)Math.ceil(Math.log(j+1)/Math.log(4))))));
    		j++;
    	}
    	System.out.println(ans);
    }
}

File: Cf/186/c.cpp
#include <iostream>
#include <stdio.h>
#include <math.h>
#include <algorithm>

using namespace std;

int main() {
    unsigned long long n;
    cin >> n;
    unsigned long long arr[n];
    unsigned long long ans = 0;
    for(int i=0; i<n; i++) {
        cin >> arr[i];
        ans += arr[i];
    }
    sort(arr, arr+n);
    int x = (int)(log(n)/log(4));
    int j=0;
    for(int i=n-1; i>=max(0ULL, n-(int)(pow(4, x-1))); i--) {
        ans += arr[i] * (x - (int)(ceil(log(j+1)/log(4))));
        j++;
    }
    cout << ans;
}
